xapian-applications/omega/omindexcheck.cc

```diff
@@ -0,0 +1,142 @@
+    tests.insert({"fb2/lang-name.fb2", {"Adavid", "Atardon", "ZAdavid",
+		  "ZFlang", "Zeel", "Zhovercraft", "Zlanguag"}});
+    tests.insert({"fb2/lang2.fb2", {"Adavid", "Atardon", "ZAtardon",
+		  "Zúhořů", "Zhovercraft", "Zlanguag", "Zmój", "Zof",
+		  "Zpełen", "Zpoduszkowiec", "Zvznášedlo", "Zwęgorzi"}});
+    tests.insert({"lrf/hello.lrf", {"ZFhello", "Zhello", "Zworld"}});
+    tests.insert({"pdb/PalmDOC-hello.pdb", {"Shello", "Sworld", "ZFpalmdoc",
+		  "Zworld"}});
+    tests.insert({"pdb/PeanutPress-hello.pdb", {"Fpeanutpress", "ZFhello",
+		  "Zhello"}});
+    tests.insert({"pdb/test.pdb", {"Sdemodemo", "ZFtest", "Zwherearew"}});
+#endif
+#if defined HAVE_LIBETONYEK
+    tests.insert({"keynotes/test-keynote.key", {"ZFkeynot", "Zbold", "Znow",
+		  "Zsubtitl"}});
+    tests.insert({"pages/test-pages.pages", {"ZFpage", "Zfull", "Zhovercraft",
+		  "Zwęgorzi"}});
+#endif
+#if defined HAVE_TESSERACT
+    tests.insert({"img/Test1.gif", {"Znoisyimag", "Zocr", "Ztesseract"}});
+    tests.insert({"img/Test2.pgm", {"ZFtest2", "Znoisyimag", "Ztesseract"}});
+    tests.insert({"img/Test3.ppm", {"ZFtest3", "Zocr", "Ztest"}});
+    tests.insert({"img/Test4.tiff", {"Znoisyimag", "Ztesseract", "Zto"}});
+    tests.insert({"img/Test5.webp", {"Znoisyimag", "Ztesseract", "Ztest"}});
+    tests.insert({"img/poster-2.jpg", {"ZFposter", "Zby", "Zproperti",
+		  "Zsurveil", "Zvideo"}});
+    tests.insert({"img/poster.jpg", {"Zbicycl", "Zride", "Zroller",
+		  "Zskateboard"}});
+    tests.insert({"img/scan-page.png", {"Zannual", "Zfed", "Zreturn",
+		  "Zwhile"}});
+#endif
+}
+
+static bool
+compare_test(testcase& test, const Xapian::Document& doc, const string& file)
+{
+    sort(test.begin(), test.end());
+    Xapian::TermIterator term_iterator = doc.termlist_begin();
+    for (auto& t : test) {
+	term_iterator.skip_to(t);
+	if (term_iterator == doc.termlist_end() || *term_iterator != t) {
+	    cerr << "Error in " << file << ": Term " << t <<
+		 " does not belong to this file" << endl;
+	    return false;
+	}
+    }
+    return true;
+}
+
+int
+main(int argc, char** argv)
+{
+    Xapian::Database db;
+    bool succeed = true;
+    if (argc <= 2)
+	return 1;
+    db.add_database(Xapian::Database(argv[2]));
+    string url, current_dir(argv[1]);
+
+    index_test();
+    for (auto t = db.allterms_begin("U"); t != db.allterms_end("U"); ++t) {
+	const string& term = *t;
+	Xapian::PostingIterator p = db.postlist_begin(term);
+	if (p == db.postlist_end(term)) {
+	    continue;
+	}
+	Xapian::docid did = *p;
+	Xapian::Document doc = db.get_document(did);
+	auto data = doc.get_data();
+	size_t start;
+	if (startswith(data, "url=")) {
+	    start = CONST_STRLEN("url=");
+	} else {
+	    start = data.find("\nurl=");
+	    if (start == string::npos)
+		continue;
+	    start += CONST_STRLEN("\nurl=");
+	}
+	url.assign(data, start, data.find('\n', start) - start);
+	start = url.find(current_dir) + current_dir.length();
+	url = url.substr(start, url.length());
```

Comment:
I would suggest simplifying this code by more carefully specifying the options to `omindex` in `omindextest`.  The only reason why need to remove `current_dir` here is because we specified it in the `--url="/$TEST_FILES"` option.  If instead we specify `--url=/` then the URL in the database will just be e.g. `/pdf/poppler.pdf`.

And with that change, we can rely on the URL term being short enough that it'll contain the full URL (if the URL term would be >= `MAX_SAFE_TERM_LENGTH` (which is 240) bytes then the end of it is hashed.

Taken together that means we can then just do `string url(term, 2);` to extract the relative path from the `U` prefix term (2 because we want to skip both the `U` and the ledding `/`) instead of having to fetch the document data and scan through it for `url=`.

Reply:
You are right, that would make the code much more readable and clear. Moreover, we are avoiding using more than `MAX_SAFE_TERM_LENGTH` characters in the `url`.